cachito/workers/scm.py

```diff
@@ -117,11 +125,22 @@ def update_and_archive(self, previous_archive):
                 log.exception("Failed to fetch from remote %s", self.url)
                 raise CachitoError("Failed to fetch from the remote Git repository")
 
-            self._reset_git_head(repo)
+            self._reset_git_head(repo, self.ref)
+            if gitsubmodule:
+                log.debug(f"Git submodules within repo are: {repo.submodules}")
+                for sm in repo.submodules:
+                    self.process_git_submodule(sm)
+
             self._create_archive(repo.working_dir)
 
-    def fetch_source(self):
-        """Fetch the repo, create a compressed tar file, and put it in long-term storage."""
+    def fetch_source(self, gitsubmodule=False):
+        """Fetch the repo, create a compressed tar file, and put it in long-term storage.
+
+        :param bool gitsubmodule: a bool to determine whether git submodules need to be processed.
+        """
+        if gitsubmodule:
+            self.sources_dir = SourcesDir(self.repo_name, f"{self.ref}-with-submodule")
```

Comment:
If we're not going to use the `-with-submodule` archives for updating (which seems reasonable), then I would suggest storing them separately from the normal archives. For example, you could add something like `submodule_archives_dir` to the `SourcesDir` class, and you would store the archives with submodules there. Example file structure:

```
.
└── some-namespace
    └── retrodep
        ├── 1735db74fe4834fb0b95cc78d5aa3e3ef1dfbb73.tar.gz
        └── with-submodules
            └── 1735db74fe4834fb0b95cc78d5aa3e3ef1dfbb73.tar.gz
```

In the `fetch_source` function, when `gitsubmodule = True`, you would then do the following:
1. If archive with submodules already exists, return
2. Fetch repo and create normal archive (without submodules) as usual
3. Process submodules and create archive with submodules

This way, the submodules workflow would be more separated from the normal git workflow, which may make things less complicated. You would just need to make sure to extract the correct archive to the `RequestBundleDir` later in `fetch_app_source`.

Reply:
Ummm, that may make things less complicated from the implementation point of view or enhance the readability but what is the other benefit of doing so over what we are doing rn except having a separate workflow?

The advantage of current workflow is that, even for a Cachtio request with submodules, in case of previous archive being present, we're not cloning the repo all over again but only adjusting the `ref` and cloning just the submodules. This makes it faster than having to clone every-time. I was trying to utilize the existing update functionality as much as possible.

Reply:
You don't have to give up the advantages of the current workflow. You would first fetch the repo using the existing mechanism (re-use existing / update previous / clone from scratch), then you would extract the archive somewhere, fetch submodules and create new archive. It could even be a separate method.

But you're right in that there are no real advantages except maybe ease of implementation/understanding, so I'm ok with keeping it as it is (but having a separate directory for archives with submodules could still be relevant)

Reply:
I took up on your suggestion and tried with a separate directory for submodules, to make it look like `<repo_name>/submodule/<ref>.tar.gz`, but there are more places than just this class where we are accessing `sources_dir.archive_path`, e.g. https://github.com/release-engineering/cachito/blob/ad3a2b94bb62607e1b328a24ba0b4b148fb769d0/cachito/workers/pkg_managers/pip.py#L1672 and so just to simplify the implementation/coding here in this class, the value of `gitsubmodule` bool would need to be propagated within multiple other classes/functions. IMO, that would just make everything else more complex. With your other suggestion, the current implementation seems to be redable and understandable and so I'd like to keep it as it is. lmk wdyt!